Working Pandora Robot Code

RobinsonZ opened the pull request:
All the code that works/was tested on Pandora before it went into the bag. Includes basic drive code, PID control for the wrist and elevator, automatic setpoints for the elevator, auto-stop intake, and turret, tape, and winch control.

This doesn't contain any auto-wrist code as that isn't quite done; that will be on another branch when this gets merged. Likewise, this doesn't contain any PID drive code as that didn't work well in testing; like wrist, that will be on another branch.

I've gone through the code and cleaned up everything that I saw; however, there are probably more efficient ways to do certain things or possible errors that I didn't see. In that sense, we should do a full review since we have time before comps.

natesales committed:
Updated ROOSTER to master and started basic robot code

natesales committed:
Updated ROOSTER implementation to master and started robot code.

Commit:
Added DriveTrain Code

Commit:
Added Power Management

Commit:
Fixed Jitpack issues

Commit:
Created AutoIntake

Commit:
Updated README with current robot controls.

Commit:
Added manual elevator controls

Commit:
Fixed terrible drivetrain control methods.

Commit:
Updated Teleop code.

Commit:
Updated motor numbers.

Commit:
Removed unused imports.

Commit:
Added auto eject functionality.

RobinsonZ committed:
Make Travis not constantly update cache

See flamingchickens1540/ROOSTER#33

Commit:
Fixed capital method names

Commit:
Added auto eject

Commit:
Added Wrist

Commit:
Updated Auto Intake

Commit:
Added Intake Variables

Commit:
Added Climber

Commit:
Updated intake with tunable speed.

Commit:
Updated intake control methods.

Commit:
Added wrist control

Commit:
Made talons private

Commit:
Updated climber with better methods

Commit:
Made wrist

Commit:
Added wrist

Commit:
Added wrist

Commit:
Moved stop intake to end

Commit:
Added manual winch control

Commit:
Removed useless line

Commit:
Added manual tape measure control

Commit:
Add multiplier for climber

Commit:
Update motor numbers

Commit:
Update controls and readme

Commit:
Fixed readme

Commit:
Removed unused imports

Commit:
Changed some Talons to Victors

Commit:
Add all joystick access methods

Commit:
Added motor test

Commit:
Add individual intake motor control

Commit:
Add Auto Climber and remove commented lines.

Commit:
Add individual winch motor control

Commit:
Add Google Java code style

Commit:
Fix power management

Commit:
Add individual motor control.

Commit:
Add servo control method

Commit:
Add leds

Commit:
Add servo control

Commit:
Add joystick elevator

Commit:
Add JoystickElevator

Commit:
Add multiple motor controllers

Commit:
Removed empty lines

RobinsonZ committed:
Merge branch 'master' into robot-base

# Conflicts:
#	build.gradle
#	src/main/java/org/team1540/robot2018/Robot.java

Commit:
Move commands into packages

Commit:
Fix broken command path

Commit:
Add elevator tuning

Commit:
Move elevator commands to package

Commit:
Add manual elevator speed tuning

Commit:
Add null option

RobinsonZ committed:
Update code style settings

RobinsonZ committed:
Merge branch 'master' into robot-base

Commit:
Clean up code

ambercmerrill committed:
made more of the robot work

RobinsonZ committed:
Optimize robot imports

That setting should be on automatically, FYI

RobinsonZ committed:
Make Intake use ChickenVictors

Since Victor SPXs don't support current limiting, disabled the
PowerManageable functionality on the intake for now. We'll need to find
a fix of some kind, maybe voltage limiting. @edelmanjm

RobinsonZ committed:
Change Intake command structure

Instead of using end() at the end of commands, uses instant commands
(through ROOSTER's SimpleCommand) with a default command to stop the
intake.

ambercmerrill committed:
fixed negations on manual intake

ambercmerrill committed:
made servos work

RobinsonZ committed:
Set tape and winch multipliers to nonzero placeholders

ambercmerrill committed:
servo disabling still doesn't work

RobinsonZ committed:
Make RunClimber command work

The isFinished() method should not be overridden.

RobinsonZ committed:
Remove useless/counterproductive overridden Command methods

Remove all overrides of interrupted() as it calls end() by default,
which is the behavior we want.

Remove other method overrides with empty bodies for obvious reasons.

RobinsonZ committed:
Split Climber into three subsystems

RobinsonZ committed:
Add missing tape and winch manual control

ambercmerrill committed:
negated winch and tape measure values

RobinsonZ committed:
Make RobotUtil use library deadzone function

RobinsonZ committed:
Remove unnecessary comment

RobinsonZ committed:
Make drivetrain control with proper axes/inversions

Commit:
Remove unused imports

RobinsonZ committed:
Switch tape inversion

We really need to be more consistent about that.

RobinsonZ committed:
Fix joystick binding

RobinsonZ committed:
Create PidTuningRobot

RobinsonZ committed:
Merge branch 'robot-base' into fixes

RobinsonZ committed:
Remove unnecessary superclass calls

Commit:
Merge pull request #6 from flamingchickens1540/fixes

Various Fixes

RobinsonZ committed:
Create HoldElevatorPosition and HoldWristPosition

We can combine these commands in command groups to accomplish any
desired movements

RobinsonZ committed:
Add missing Tuning fields

RobinsonZ committed:
Replace PidTuningRobot with robot just for wrist

RobinsonZ committed:
Merge branch 'robot-base' into pid

# Conflicts:
#	src/main/java/org/team1540/robot2018/subsystems/Elevator.java
#	src/main/java/org/team1540/robot2018/subsystems/Wrist.java

RobinsonZ committed:
Change elevator to invert properly

Also fix naming issues

RobinsonZ committed:
Add superstructure command groups

Commit:
Merge pull request #7 from flamingchickens1540/fixes

More fixes

RobinsonZ committed:
Reset default commands to joystick control

RobinsonZ committed:
Add peak outputs and ramp rates

RobinsonZ committed:
Add exchange command group

RobinsonZ committed:
Add proper tuned values

RobinsonZ committed:
Add additional test logic to prevent oscillation around a setpoint

RobinsonZ committed:
Merge branch 'robot-base' into pid

RobinsonZ committed:
Fix missing tuning constant

RobinsonZ committed:
Implement Motion Magic control for elevator

RobinsonZ committed:
Add button bindings

Also, disable wrist

RobinsonZ committed:
Change JoystickElevator to be non-PID backup control

RobinsonZ committed:
Replace isOutsideRange() logic with ROOSTER library function

RobinsonZ committed:
Bind most OI controls

Also, refactor buttons to things that make sense

RobinsonZ committed:
Improve elevator and wrist control logic

Now avoids hitting the center bar when in base position commands.

Commit:
Move wrist out of intake package

Commit:
Add updatePID

Commit:
Update PID values

Commit:
Change elevator to wrist

Commit:
Update wrist PID control

RobinsonZ committed:
Replace FrontScale wrist logic

RobinsonZ committed:
Add obstacle avoidance and wrist logic

Commit:
Refactor wrist names

RobinsonZ committed:
Merge branch 'robot-base' into pid

# Conflicts:
#	src/main/java/org/team1540/robot2018/Tuning.java
#	src/main/java/org/team1540/robot2018/commands/elevator/JoystickElevator.java
#	src/main/java/org/team1540/robot2018/subsystems/Wrist.java

RobinsonZ committed:
Move PID updating into periodic methods

RobinsonZ committed:
Merge remote-tracking branch 'origin/pid' into pid

RobinsonZ committed:
Fix misc controls issues

There are still several things that don't work that should.

RobinsonZ committed:
Make wrist and elevator functional again

RobinsonZ committed:
Add maximum time to auto-intake

RobinsonZ committed:
Invert elevator

RobinsonZ committed:
Add correct wrist positions

RobinsonZ committed:
Add wrist joystick control

RobinsonZ committed:
Improve controls

RobinsonZ committed:
Add pandora/titan switching

For certain specific things like PDP ports for the intake current.

RobinsonZ committed:
Disable tape for now

RobinsonZ committed:
Add camera

RobinsonZ committed:
Remove motor 0

RobinsonZ committed:
Change climber constants

RobinsonZ committed:
Make intake run in only after auto-intake succeeds

RobinsonZ committed:
Add current telemetry to elevator

RobinsonZ committed:
Rebind climber controls

RobinsonZ committed:
Add relays

RobinsonZ committed:
Add per-robot elevator inversion

RobinsonZ committed:
Make relay enable and disable when changing robot modes

RobinsonZ committed:
Remove PowerManageable implementation from wrist

RobinsonZ committed:
Improve idle intake grab handling

RobinsonZ committed:
Periodically set elevator inversions

RobinsonZ committed:
Add default values for peak output forward and reverse

RobinsonZ committed:
Increase manual control deadzone

RobinsonZ committed:
Add wrist current safety

RobinsonZ committed:
Remove auto wrist flip after intake

Change was actually made in 9aeea02

RobinsonZ committed:
Reorganize Tuning

- Make all names in line with style guide. None of them are constants
(they're non-final) and also seriously why UpperCamelCase for fields

- Certain fields have no effect after the robot is started up, so took
those off SmartDashboard to keep it less cluttered

- Organized field order slightly

- Deleted unused fields

- Accelleration -> Acceleration

RobinsonZ committed:
Replace all processAxisDeadzone() calls

RobinsonZ committed:
Update GradleRIO version

RobinsonZ committed:
Merge branch 'robot-base' into pid-nosep

RobinsonZ committed:
Create DriveTestRobot

RobinsonZ committed:
Integrate ROOSTER PID drive functionality

RobinsonZ committed:
Integrate ROOSTER motion profiling

RobinsonZ committed:
Fix final fields

RobinsonZ committed:
Add brake inversion for Pandora

RobinsonZ committed:
Fix elevator inversions and make them update properly

RobinsonZ committed:
Fix tuning variable typo

RobinsonZ committed:
Check for timeout when ending AutoIntake

Originally distinguished between being canceled and intaking a cube via
interrupted() and end() calls. However, end() could also be called when
timing out resulting in the motors spinning inwards and looking a bit
stupid. Now end() checks if the command timed out before setting the
motors to grab.

ambercmerrill committed:
Fix a lot of things

ambercmerrill committed:
Negate some values so it works on Pandora, also change some of the copilot controls

RobinsonZ committed:
Remove unused manual climber controls

They've been replaced by control definitions in Robot

RobinsonZ committed:
Refactor intake bindings

Unnecessary to have a button that cancels the intake command; it can
just start another command with the intake as a requirement.

RobinsonZ committed:
Remove blocks of commented drive code

RobinsonZ committed:
Remove MoveWristToPosition

This also removes the elevator safety logic. Wrist commands will be
added back in in a separate branch; however, they were not functional
when Pandora went in the bag so they aren't going in master yet.

RobinsonZ committed:
Remove unused intake commands

HoldIntake's functionality is performed at the end of AutoIntake
now.

RobinsonZ committed:
Rename AutoEject to EjectCube

Nothing automatic/sensor based about it and creates a confusing
dichotomy with AutoIntake.

RobinsonZ committed:
Add stop() method to tape measure

This allows for a method reference to be passed more concisely instead
of a lambda.

RobinsonZ committed:
Extract default pan and tilt to Tuning fields

RobinsonZ committed:
Add stop() method to winch

See similar commit a4405e7; allows for increased conciseness creating
SimpleCommands.

RobinsonZ committed:
Refactor intake

Friends don't let friends use snake case.

RobinsonZ committed:
Remove LED subsystem

Untested and not reflective of the actual robot.

RobinsonZ committed:
Clean up Robot

- Remove commented bindings

- Remove camera (will be re-added with fancy crosshair etc.)

- Remove unnecessary SmartDashboard telemetry

RobinsonZ committed:
Refactor RobotMap constants

Now in line with Google Style constant naming standards.

RobinsonZ committed:
Refactor Tuning constants

Two last ones I missed.

edelmanjm commented:
oh jesus christ

edelmanjm commented on README.md:
```diff
@@ -1,3 +1,8 @@
+
+#### Copilot
+![Copilot Controls](img/copilot_controller.png)
+#### Driver:
```
Inconsistent colons

RobinsonZ commented:
Don't worry, about ~a third~ 20% of the line count is IntelliJ config XML files which don't need review.

edelmanjm commented on src/main/java/org/team1540/robot2018/DriveTestRobot.java:
```diff
@@ -0,0 +1,184 @@
+package org.team1540.robot2018;
```
Test class, I'm not reviewing this

edelmanjm commented on src/main/java/org/team1540/robot2018/ElevatorTuningRobot.java:
```diff
@@ -0,0 +1,192 @@
+package org.team1540.robot2018;
```
Also not reviewing this

edelmanjm commented on src/main/java/org/team1540/robot2018/OI.java:
```diff
@@ -0,0 +1,160 @@
+  public static Joystick driver = new Joystick(0);
+  public static Joystick copilot = new Joystick(1);
+
+  public static final int X = 3;
```
Not following naming conventions?

RobinsonZ replied:
Static final immutable fields use constant case per [Google Style](https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names).

edelmanjm replied:
What about `back` and `start` then?

RobinsonZ replied:
Fair point, will refactor those.

edelmanjm commented on src/main/java/org/team1540/robot2018/OI.java:
```diff
@@ -0,0 +1,160 @@
+
+public class OI {
+
+  public static Joystick driver = new Joystick(0);
```
Why do these need to be public?

RobinsonZ replied:
One usage in `MotorTestRobot` but you're right, that can be made private, and `MotorTestRobot` shouldn't depend on `OI` anyway.

edelmanjm commented on src/main/java/org/team1540/robot2018/OI.java:
```diff
@@ -0,0 +1,160 @@
+  // manual tape in
+  static Button copilotStart = new JoystickButton(copilot, start);
+
+  static Button elevatorJoystickActivation = new Button() {
```
Can't these be replaced by lambdas?

edelmanjm replied:
Also if you're going to be creating these en mass isn't there a cleaner way using a builder or something

RobinsonZ replied:
As they're abstract classes, no, they can't be replaced by lambdas

RobinsonZ replied:
A button class that takes a `BooleanSupplier` would be a good addition to ROOSTER, though.

edelmanjm commented on src/main/java/org/team1540/robot2018/OI.java:
```diff
@@ -0,0 +1,160 @@
+    }
+  };
+
+  // static Button copilotLeftTriggerSmallPress = new Button() {
```
Commented blocks in master? Booo

RobinsonZ replied:
Whoops, missed that.

RobinsonZ commented on README.md:
```diff
@@ -1,3 +1,8 @@
 Team 1540's robot-specific code for our 2018 robots, Pandora and Titan. Does not include libraries reused from year-to-year; those can be found in [flamingchickens1540/ROOSTER](https://github.com/flamingchickens1540/ROOSTER).
+
+#### Copilot
+![Copilot Controls](img/copilot_controller.png)
```
Should we make any effort to keep that image up-to-date or just remove it until we have final controls?

ambercmerrill commented:
the (very outdated) version of the copilot controls needs to be removed. (the more updated google doc is also out of date but we can update that later)

edelmanjm commented on src/main/java/org/team1540/robot2018/Robot.java:
```diff
@@ -1,17 +1,82 @@
+import org.team1540.robot2018.commands.groups.IntakeSequence;
+import org.team1540.robot2018.commands.intake.EjectCube;
+import org.team1540.robot2018.commands.wrist.JoystickWrist;
+import org.team1540.robot2018.subsystems.ClimberTapeMeasure;
```
.*?

RobinsonZ replied:
[Nope.](https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports)

RobinsonZ committed:
Remove outdated control diagrams

RobinsonZ commented:
@ambercmerrill Done.

edelmanjm commented on src/main/java/org/team1540/robot2018/Robot.java:
```diff
@@ -1,17 +1,82 @@
+
+    OI.elevatorJoystickActivation.whileHeld(new JoystickElevator());
+
+    OI.wristJoystickActivation.whileHeld(new ConditionalCommand(new AlignClimber(), new JoystickWrist()) {
```
I can follow, but this still deserves a comment

RobinsonZ commented:
Oh wait, I need to actually remove the image.

RobinsonZ committed:
Remove copilot controller image

Since I forgot the first time around.

edelmanjm commented on src/main/java/org/team1540/robot2018/RobotMotorTest.java:
```diff
@@ -0,0 +1,166 @@
+package org.team1540.robot2018;
```
Not reviewing this

edelmanjm commented on src/main/java/org/team1540/robot2018/Tuning.java:
```diff
@@ -1,4 +1,201 @@
 
+import org.team1540.base.adjustables.Tunable;
+
 public class Tuning {
```
Trying to do anything with everything just dumped in SmartDashboard is horrifying difficult. These absolutely should be grouped in some hierarchical fashion. Sendables exist for a reason.

RobinsonZ replied:
Fair, but adding that functionality to the `AdjustableManager` in a sane way is a fairly large undertaking. If you want to not use the manager and just set up `Tuning` to have a `Sendable` table, be my guest, but that's a bunch of uber-boilerplate that I don't want to deal with.

edelmanjm replied:
Needs to be done in some capacity though, please. Trying to work with this as-is is a total PITB. I'll do this if I have time I guess.

edelmanjm commented on src/main/java/org/team1540/robot2018/commands/JoystickDrive.java:
```diff
@@ -0,0 +1,23 @@
+  @Override
+  protected void execute() {
+    double triggerValue = OI.getDriverRightTrigger() - OI.getDriverLeftTrigger();
+    Robot.drivetrain.setLeft(-OI.getDriverLeftY() + triggerValue);
```
Should the negation be done here on in `getDriver_()`?

RobinsonZ replied:
Here. Randomly negating things in `OI` is just confusing.

RobinsonZ committed:
Run full auto-format

edelmanjm commented on src/main/java/org/team1540/robot2018/commands/climber/AlignClimber.java:
```diff
@@ -0,0 +1,32 @@
+
+  @Override
+  protected boolean isFinished() {
+    return false; //Return true to stop the command
```
Shouldn't there be a flag and an accessor?

RobinsonZ replied:
What for? Like most/all joystick control commands, it's intended to only cancel when another command interrupts it.

edelmanjm commented on src/main/java/org/team1540/robot2018/commands/climber/WinchOut.java:
```diff
@@ -0,0 +1,27 @@
+
+  @Override
+  protected boolean isFinished() {
+    return false; //Return true to stop the command
```
Again, flag and accessor?

RobinsonZ replied:
`WinchOut` may actually be an issue as I believe that should be returning `true` since it's continuously called. However, AFAIK we're removing this entirely.

edelmanjm commented on src/main/java/org/team1540/robot2018/commands/elevator/HoldElevatorPosition.java:
```diff
@@ -0,0 +1,31 @@
+import org.team1540.robot2018.Tuning;
+
+public class HoldElevatorPosition extends Command {
+  double setpoint;
```
Private?

edelmanjm commented on src/main/java/org/team1540/robot2018/commands/elevator/HoldElevatorPosition.java:
```diff
@@ -0,0 +1,31 @@
+
+  @Override
+  protected void execute() {
+    if (Math.abs(Robot.elevator.getPosition() - setpoint) > Tuning.maxElevatorDeviation) {
```
A comment would be nice

edelmanjm commented on src/main/java/org/team1540/robot2018/commands/elevator/JoystickElevator.java:
```diff
@@ -0,0 +1,22 @@
+
+  @Override
+  protected void execute() {
+    Robot.elevator.set(-OI.getCopilotLeftY());
```
Again, should negation be done in OI?

RobinsonZ replied:
Nope. See above.

edelmanjm commented on src/main/java/org/team1540/robot2018/commands/elevator/MoveElevatorToPosition.java:
```diff
@@ -0,0 +1,10 @@
+public class MoveElevatorToPosition extends CommandGroup {
+
+  public MoveElevatorToPosition(double target) {
+    addSequential(new MoveElevator(target));
```
...wait a one command CommandGroup? Why? If there's a reason add a TODO

RobinsonZ replied:
That used to (and will soon) have wrist safety logic; I'll add a TODO for now.

ambercmerrill commented on src/main/java/org/team1540/robot2018/Tuning.java:
```diff
@@ -1,4 +1,201 @@
+  public static double winchInSpeed = 1;
+
+  @Tunable("Winch Out Speed")
+  public static double winchOutSpeed = 1; // 0.75
```
the winch in and out speeds shouldn't both be positive 1. it doesn't actually matter which is which direction (because we're using the stuff sack box) but the out speed should be negated.

ambercmerrill commented on src/main/java/org/team1540/robot2018/Tuning.java:
```diff
@@ -1,4 +1,201 @@
+  @Tunable("Climber In Low Speed")
+  public static double climberInLowSpeed = -0.4;
+  @Tunable("Climber In HIgh Speed")
+  public static double climberInHighSpeed = -1;
```
all the climber constants in Tuning need to be cleaned up. check which constants are actually getting used, because a lot of them need to be deleted, and also they should all be in the same place.

ambercmerrill commented on src/main/java/org/team1540/robot2018/commands/climber/WinchOut.java:
```diff
@@ -0,0 +1,27 @@
+
+public class WinchOut extends Command {
+
+  public WinchOut() {
```
are any of these classes (WinchOut, WinchIn, TapeOut, TapeIn, etc, i know there's more) actually getting used? i think it's better style to actually use them, but if we aren't doing that, delete them all.

RobinsonZ replied:
`WinchOut` is used [here](https://github.com/flamingchickens1540/pandora-titan-2018/pull/8/files#diff-ee56816ca2f61bdffc0991b1df2b8b8fR43). All the others are already deleted.

RobinsonZ replied:
Of course, `WinchOut` could easily be replaced with a `SimpleCommand` since that appears to be what we're doing anyway.

edelmanjm commented on src/main/java/org/team1540/robot2018/commands/wrist/HoldWristPosition.java:
```diff
@@ -0,0 +1,42 @@
+import org.team1540.robot2018.Tuning;
+
+public class HoldWristPosition extends Command {
+  double setpoint;
```
Again private?

edelmanjm commented on src/main/java/org/team1540/robot2018/commands/wrist/HoldWristPosition.java:
```diff
@@ -0,0 +1,42 @@
+public class HoldWristPosition extends Command {
+  double setpoint;
+  private long lastExecTime;
+  private long spikeDuration;
```
What exactly is `spikeDuration` being used for, and why not a timer?

RobinsonZ replied:
Fair point; a timer would be a better idea.

edelmanjm commented on src/main/java/org/team1540/robot2018/subsystems/ClimberTapeMeasure.java:
```diff
@@ -0,0 +1,23 @@
+import org.team1540.robot2018.RobotMap;
+
+public class ClimberTapeMeasure extends Subsystem {
+  private ChickenVictor victor = new ChickenVictor(RobotMap.TAPE_MEASURE);
```
nice naming :P

edelmanjm commented on src/main/java/org/team1540/robot2018/subsystems/ClimberWinch.java:
```diff
@@ -0,0 +1,41 @@
+import org.team1540.robot2018.RobotMap;
+
+/**
+ * MADE_OF_ALUMINIUM -> IS_WINCH
```
wat

RobinsonZ replied:
bad csIII joke. ignore.

edelmanjm commented on src/main/java/org/team1540/robot2018/subsystems/ClimberWinch.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * MADE_OF_ALUMINIUM -> IS_WINCH
+ */
+public class ClimberWinch extends Subsystem {
```
so like at some point ChickenSubsystem amirite guys

edelmanjm commented on src/main/java/org/team1540/robot2018/subsystems/DriveTrain.java:
```diff
@@ -0,0 +1,105 @@
+  private ChickenTalon[] masters = new ChickenTalon[]{left, right};
+
+  public DriveTrain() {
+    this.add(talons);
```
You can just do add(...)

edelmanjm commented on src/main/java/org/team1540/robot2018/subsystems/DriveTrain.java:
```diff
@@ -0,0 +1,105 @@
+
+  @Override
+  public void periodic() {
+    for (ChickenTalon talon : masters) {
```
Are these really changing so often they need to be constantly set?

RobinsonZ replied:
Little to no overhead, and allows them to be tuned on the fly.

edelmanjm replied:
Why not just have an update method that gets called when your tunables are modified?

RobinsonZ replied:
You'd have to do constant edge detection since you can't really have proper listeners due to the way NetworkTables (and by extension SmartDashboard and by extension the `AdjustableManager`) works. At that point, it's a good bit more code for a negligible (maybe nonexistent or negative if CTRE's software checks if values are the same before sending them) performance benefit.

edelmanjm replied:
Maybe I'm missing something on how NetworkTables work, but couldn't you just trip a flag in the NetworkTable and then run all the updating? Then you're only constantly checking one value.

edelmanjm replied:
Also you could spin that off into another thread and have it sleep a bunch.

edelmanjm replied:
Although then locking ew.

RobinsonZ replied:
Nope. The `AdjustableManager` is literally just an annotation-based wrapper around the `SmartDashboard.putX` and `SmartDashboard.getX` functions. Having a boolean flag set when a value is modified would basically entail writing our own SmartDashboard that supported that sort of thing, because SmartDashboard right now is just a dressed-up Outline Viewer with graphs.

edelmanjm replied:
If you decide tho to do grouping for all the organization stuff (earlier comment) anyways, then you've got a convenient group with which to use your flag.

RobinsonZ replied:
Fair, but that would also require changes on the library level to be effective and I'm not sure if SmartDashboard supports "modified" flags anyway.

edelmanjm commented on src/main/java/org/team1540/robot2018/subsystems/DriveTrain.java:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  public double getLeftPosition() {
+    return left.getSelectedSensorPosition();
```
You can also directly grab the quadrature since we don't seem to do anything else anyways

RobinsonZ replied:
Why not be polymorphic?

edelmanjm replied:
Accidents with grabbing the wrong sensor mostly if someone accidentally sets something; if we know that it needs to be from the quad, get it from the quad. Although sure that's fine if that's what we want.

RobinsonZ replied:
There's nothing software-wise _preventing_ it from being a different type of encoder, you'd just plug different values in. While defensive programming is a good thing generally, the feedback sensor can only be set within the same source file and if feedback sensors are getting set willy-nilly we have a much bigger problem.

edelmanjm commented on src/main/java/org/team1540/robot2018/subsystems/Elevator.java:
```diff
@@ -0,0 +1,76 @@
+    talon1.configSelectedFeedbackSensor(FeedbackDevice.QuadEncoder);
+    Command command = new SimpleCommand("Zero Elevator", () -> talon1.setSelectedSensorPosition(0));
+    command.setRunWhenDisabled(true);
+    SmartDashboard.putData(command);
```
uh why? Also adding stuff into SmartDashboard in a decentralized manner

RobinsonZ replied:
Allows zeroing of the elevator if it's out of alignment, whether due to just being turned on or something breaking.

edelmanjm replied:
Centralize it then?

RobinsonZ replied:
Can do, I suppose.

edelmanjm commented on src/main/java/org/team1540/robot2018/subsystems/Elevator.java:
```diff
@@ -0,0 +1,76 @@
+
+  @Override
+  public void periodic() {
+    talon1.config_kP(0, Tuning.elevatorP);
```
Again are these changing so often they need to be in periodic?

RobinsonZ replied:
They aren't changing often, but when they do, they need to be updated.

edelmanjm commented on src/main/java/org/team1540/robot2018/subsystems/Intake.java:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  public double getCurrent() {
+    return (Tuning.isPandora ? pdp.getCurrent(5) : pdp.getCurrent(10) + pdp.getCurrent(11))
```
lmao

RobinsonZ replied:
We can't get current telemetry from victors, so I consider this innovation.

edelmanjm commented on src/main/java/org/team1540/robot2018/subsystems/Wrist.java:
```diff
@@ -0,0 +1,80 @@
+    SmartDashboard.putData(command);
+  }
+
+  public double getTrajPosition() {
```
non no abbreviations

edelmanjm commented on src/main/java/org/team1540/robot2018/subsystems/Wrist.java:
```diff
@@ -0,0 +1,80 @@
+
+  @Override
+  public void periodic() {
+    // System.out.println("periodic");
```
hrm

edelmanjm commented on src/main/java/org/team1540/robot2018/subsystems/Wrist.java:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  public double getPosition() {
+    return wristMotor.getSelectedSensorPosition();
```
again do we wanna just access the quad directly?

RobinsonZ replied:
Nope.

edelmanjm commented on src/main/java/org/team1540/robot2018/subsystems/Wrist.java:
```diff
@@ -0,0 +1,80 @@
+  @Override
+  public void periodic() {
+    // System.out.println("periodic");
+    wristMotor.config_kP(0, Tuning.wristP);
```
Also again does this all really need to be periodic?

RobinsonZ replied:
Yup.

edelmanjm reviewed (changes requested):
Mostly just Tuning, the rest is stuff too but not as annoying

edelmanjm commented:
oh dear god i just got the little watch thing reviewing this

RobinsonZ committed:
Make OI joysticks private

Per @edelmanjm's review

RobinsonZ committed:
Refactor OI button constants

RobinsonZ committed:
Remove unused climber tuning constants

RobinsonZ committed:
Add explanatory comment to wristJoystickActivation

[ci skip]

RobinsonZ committed:
Improve HoldElevatorPosition niceness

Make a field private that I forgot about, and add an explanatory
comment.

RobinsonZ committed:
Add TODO to MoveElevatorToPosition

RobinsonZ committed:
Improve HoldWristPosition

Replace weird timing system with WPILib Timer class.

Also, make a field private.

RobinsonZ committed:
Remove unnecessary this qualifiers

RobinsonZ committed:
Remove commented block

RobinsonZ committed:
Improve Wrist code style

Remove commented debug print statement, and expand an abbreviation.

RobinsonZ committed:
Move all SmartDashboard initialization to Robot

RobinsonZ committed:
Remove WinchOut and associated bindings

ambercmerrill commented on src/main/java/org/team1540/robot2018/Tuning.java:
```diff
@@ -1,4 +1,193 @@
+
+  // WINCH
+  @Tunable("Winch Out Speed")
+  public static double winchOutSpeed = 1; // 0.75
```
rename to winch in (winch never goes out)

RobinsonZ replied:
This is actually never used, so I’m just deleting it.

ambercmerrill commented on src/main/java/org/team1540/robot2018/Tuning.java:
```diff
@@ -1,4 +1,193 @@
+  public static double tapeMeasureMultiplier = 1;
+
+  @Tunable("Winch Multiplier")
+  public static double winchMultiplier = 1;
```
look at the climber related constants that are actually getting used. the only tape/winch constants we should be using are:
- winch in
- tape out
- tape in (this one synced to winch in)
and the "climber" in and out speeds below should be deleted or renamed.

ambercmerrill commented on src/main/java/org/team1540/robot2018/Tuning.java:
```diff
@@ -1,4 +1,193 @@
+
+  public static double drivetrainEncoderTPU;
+
+  // TURRET
```
put this section up with the rest of the climber

RobinsonZ committed:
Reconfigure climber constants

Completely separated tape/winch speeds (instead of having some kind of
multiplier). May need additional tape in speed for manual retraction,
pending testing.

Also, group turret and winch/tape constants in the file.

natesales reviewed (approved):
I wasn't here for some testing of this, code looks fine to me. Why isn't .idea/ gitignored?

RobinsonZ commented:
@nwsnate see flamingchickens1540/ROOSTER#5
